src/misc/msccl/msccl_lifecycle.cc

```diff
@@ -250,7 +228,7 @@ ncclResult_t mscclInit(ncclComm_t comm) {
     status.lastStream = nullptr;
     NCCLCHECK(mscclInitWorkFifoStatus(&(status.defaultWorkFifoStatus)));
 
-    mscclInitialized.store(true, std::memory_order_release);
+    mscclInitialized(comm->rank) = true;
```

Comment:
It would be better to have a function that dedicated for setting initialized to true. With that, we will able to implement lock to avoid race conditions and detect programming error like initializing msccl twice on same rank.

Reply:
I can implement as a function, and detect the programming error. However, a lock isn't necessary here because this flag is only accessed from a single rank. Thoughts?

Reply:
I think the lock is to prevent race condition between read/write from different threads. BTW, there is an inherent bug from current MSCCL API: even in one process per GPU mode, each GPU can be associated with multiple communicators and has different rank numbers. mscclLoadAlgo(... int rank) was designed incorrectly. Maybe we should deprecate this API as well, and always use comm as mapping key

Reply:
Only one thread should ever be accessing that flag at one time.

We should chat about the MSCCL bug. It might be out of scope for this work, or maybe we can easily deal with it here.

Reply:
sure. at least the patch will not make it worse than what we have today. We can address the original issue later